src/com/ms/silverking/cloud/dht/daemon/storage/FileSegment.java

```diff
@@ -23,243 +23,255 @@
-        
-        dataBuf = raFile.getChannel().map(MapMode.READ_WRITE, 0, dataSegmentSize);
-        dataBuf.put(header);
-        return new FileSegment(nsDir, segmentNumber, AccessMode.Creation, raFile, dataBuf, dataSegmentSize, nsOptions);
-    }
-    
-    public static FileSegment openForDataUpdate(File nsDir, int segmentNumber, int dataSegmentSize,
-                                                SyncMode syncMode, NamespaceOptions nsOptions, 
-                                                SegmentIndexLocation segmentIndexLocation, SegmentPrereadMode segmentPrereadMode) throws IOException {
-        return open(nsDir, segmentNumber, AccessMode.Update, dataSegmentSize, syncMode, nsOptions, segmentIndexLocation, segmentPrereadMode);
-    }
-    
-    public static FileSegment openReadIndexOnly(File nsDir, int segmentNumber, int dataSegmentSize, 
-            NamespaceOptions nsOptions) throws IOException {
-        return open(nsDir, segmentNumber, AccessMode.ReadIndexOnly, dataSegmentSize, SyncMode.NoSync, nsOptions, SegmentIndexLocation.RAM, SegmentPrereadMode.NoPreread);
-    }
-    
-    public static FileSegment openReadOnly(File nsDir, int segmentNumber, int dataSegmentSize, NamespaceOptions nsOptions) throws IOException {
-        return openReadOnly(nsDir, segmentNumber, dataSegmentSize, nsOptions, SegmentIndexLocation.RAM, SegmentPrereadMode.Preread);
+  private final AccessMode accessMode;
+  private RandomAccessFile raFile;
+  private final FileSegmentStorageFormat storageFormat;
+
+  private static final String roFileMode = "r";
+  private static final String rwFileMode = "rw";
+  private static final String rwdFileMode = "rwd";
+
+  enum AccessMode {
+    Creation,     // Creation of segment. Only allowed once per segment
+    Repair,       // Repair a segment with a damaged index; recovery only
+    Update,       // Update segment entries, but not index
+    ReadOnly,     // Read-only access to segment
+    ReadIndexOnly // Read-only access to the segment index only; data proper is not mapped
+  }
+
+  ;
+
+  enum SyncMode {NoSync, Sync}
+
+  ;
+
+  enum SegmentPrereadMode {NoPreread, Preread}
+
+  ;
+
+  private static final boolean debugMetaData = false;
+
+  private static String syncModeToFileOpenMode(SyncMode syncMode) {
+    return syncMode == SyncMode.NoSync ? rwFileMode : rwdFileMode;
+  }
+
+  static final FileSegment create(File nsDir, int segmentNumber, int dataSegmentSize, SyncMode syncMode,
+      NamespaceOptions nsOptions) throws IOException {
+    RandomAccessFile raFile;
+    byte[] header;
+    ByteBuffer dataBuf;
+    int indexOffset;
+
+    indexOffset = dataSegmentSize;
+    raFile = new RandomAccessFile(fileForSegment(nsDir, segmentNumber), syncModeToFileOpenMode(syncMode));
+    header = SegmentFormat.newHeader(segmentNumber, dataOffset, indexOffset);
+
+    dataBuf = raFile.getChannel().map(MapMode.READ_WRITE, 0, dataSegmentSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getChannel*. Currently, there are execution paths that do not contain closure statements, for example, when FileChannel.map throws an exception. Either a) close the object returned by *getChannel()* in a try-finally block or b) close the resource by declaring the object returned by *getChannel()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).